### PR TITLE
Document the remote driver protocol

### PR DESCRIPTION
**Pull request by squaremo**
And some minor updates to the preceding text.

Signed-off-by: Michael Bridgen <mikeb@squaremobius.net>

### CODE REVIEW

File: docs/remote.md [hunk 1]
```diff
@@ -1,18 +1,191 @@
...
+            "Address": string,
+            "AddressIPv6": string,
+            "MacAddress": string
+        }, ...]
+    }
+
+with values in the `Interfaces` entries as above. For each entry, an `ID` and `MacAddress` and one of `Address` and `AddressIPv6` must be given. The `ID` is arbitrary but must differ among entries. It is used to identify, within the scope of the endpoint, an individual interface during a `Join` call.
+
+If the remote process was supplied entries in `Interfaces`, it must response with an empty `Interfaces` list.
+
```
Comment (mavenugo) [inline docs/remote.md]: It would be good to be explicit and mention the fact that in this case if the driver returns an `Interfaces` list, it is considered as an error and it will result in a rollback operation.

File: docs/remote.md [hunk 2]
```diff
@@ -1,18 +1,191 @@
...
+        "Gateway": string,
+        "GatewayIPv6": string,
+        "HostsPath": string,
+        "ResolvConfPath": string
+    }
+
+`Gateway` is optional and if supplied is an IP address as a string; e.g., `"192.168.0.1"`. `GatewayIPv6` is optional and if supplied is an IPv6 address as a string; e.g., `"fe80::7809:baff:fec6:7744"`. `HostsPath` is optional, as is `ResolvConfPath`.
+
+The entries in `InterfaceNames` represent veths that should be moved by LibNetwork into the sandbox; the `SrcName` is the name of the veth that the remote process created, and the `DstName` is the name the veth should have after it has been moved into the sandbox.
+
```
Comment (mavenugo) [inline docs/remote.md]: I think this changed to DstPrefix ?

File: docs/remote.md [hunk 3]
```diff
@@ -1,18 +1,191 @@
...
+            "AddressIPv6": string,
+            "MacAddress": string
+        }, ...]
+    }
+
+The `NetworkID` is the generated identifier for the network to which the endpoint belongs; the `EndpointID` is a generated identifier for the endpoint.
+
+`Options` is an arbitrary map as supplied to the proxy.
+
+The `Interfaces` value is a list with values of the form given. The fields in the `Interfaces` entries may be empty; and the `Interfaces` list itself may be empty. If supplied, `Address` is an IPv4 address and subnet in CIDR notation; e.g., `"192.168.34.12/16"`. If supplied, `AddressIPv6` is an IPv6 address and subnet in CIDR notation. `MacAddress` is a MAC address as a string; e.g., `"6e:75:32:60:44:c9"`.
```
Comment (spikecurtis) [inline docs/remote.md]: It would be helpful to explicitly describe under what circumstances the `Interfaces` list will be non-empty, and what the driver is expected to do (e.g. not just what form the information takes).

E.g. If the end user requests a specific IP address to be assigned to an endpoint, the Interface list will contain an interface with that IP address.  If the interface is provided, the driver should attempt to assign the requested IP addresses from it's IPAM subsystem.  If the requested IP address cannot be allocated, the driver should return an error.

Or whatever is actually the case!

**Comment by mavenugo** [comment]
@squaremo apologies on the delayed review. Thanks for a good writeup. couple of minor comments. Can you please take care of them ?

File: docs/remote.md (hunk 3 above)
Comment (squaremo) [inline docs/remote.md]: > It would be helpful to explicitly describe under what circumstances the Interfaces list will be non-empty, and what the driver is expected to do

I'd like to know that too! It is still mysterious -- see for instance the discussion at https://github.com/docker/libnetwork/pull/162#discussion_r30647017

**Comment by squaremo** [comment]
Rebased (probably unnecessary) and updated to account for DstPrefix and static routes.

File: docs/remote.md (hunk 3 above)
Comment (mavenugo) [inline docs/remote.md]: @squaremo  the use-case that @spikecurtis mentioned above is a valid use-case of such an instance and can be used as an example. And the expected behavior of the driver is to honor that request. If it cannot, it should return an error (Exactly like @spikecurtis suggested).

Comment (squaremo) [inline docs/remote.md]: Where are these semantics documented?

Comment (mavenugo) [inline docs/remote.md]: Its kinda documented in driverapi.go on `CreateEndpoint`.

```
> // The endpoint information can be either consumed by
> // the driver or populated by the driver.
```
But, I thought this PR is more appropriate to explain it in person, this being a Documentation PR ?

Comment (squaremo) [inline docs/remote.md]: > Its kinda documented in driverapi.go

"kinda" is accurate -- that doesn't say anything about the driver treating a non-empty list as an obligation, for instance.

> But, I thought this PR is more appropriate to explain it in person, this being a Documentation PR ?

This file is about the _remote_ protocol, and describes the constraints on the messages back and forth. The semantics of driverapi methods apply to _all_ drivers, not just remote drivers. There's no problem referring to, or recapitulating, semantics detailed elsewhere though.

Comment (mavenugo) [inline docs/remote.md]: you are right. if you feel strongly about it, can you update the doc on driverapi.go to reflect this semantics ?

Comment (squaremo) [inline docs/remote.md]: > can you update the doc on driverapi.go to reflect this semantics ?

Sure, why don't I do that in another PR, and simply say here "for the semantics of these RPCs, see the driver documentation"?

Comment (mavenugo) [inline docs/remote.md]: works for me.

**Comment by mavenugo** [comment]
Pls squash the commits.

**Comment by squaremo** [comment]
> Pls squash the commits.

I've added a link to the general design doc for driver semantics, and squashed all commits.

**Comment by mavenugo** [comment]
Thanks. LGTM

File: docs/remote.md (hunk 3 above)
Comment (squaremo) [inline docs/remote.md]: https://github.com/docker/libnetwork/pull/260

**Comment by mrjana** [comment]
LGTM
